File: pylex/lexer.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto
from typing import List, Optional, Tuple, Union


class LexerError(Exception):
    def __init__(self, message: str, line: int, column: int):
        super().__init__(f"{message} at line {line}, column {column}")
        self.message = message
        self.line = line
        self.column = column


class TokenType(Enum):
    IDENTIFIER = auto()
    KEYWORD = auto()

    INT = auto()
    FLOAT = auto()
    STRING = auto()

    OP = auto()  # Operators like ==, !=, <=, >=, +, -, **, //, ...
    PUNCTUATION = auto()  # Punctuation like (, ), {, }, [, ], :, ,, .

    NEWLINE = auto()
    INDENT = auto()
    DEDENT = auto()

    EOF = auto()


@dataclass
class Token:
    type: TokenType
    value: Optional[Union[str, int, float]]
    line: int
    column: int

    def __repr__(self) -> str:
        return f"Token(type={self.type}, value={self.value!r}, line={self.line}, column={self.column})"


KEYWORDS = {
    "False",
    "None",
    "True",
    "and",
    "as",
    "assert",
    "async",
    "await",
    "break",
    "class",
    "continue",
    "def",
    "del",
    "elif",
    "else",
    "except",
    "finally",
    "for",
    "from",
    "global",
    "if",
    "import",
    "in",
    "is",
    "lambda",
    "nonlocal",
    "not",
    "or",
    "pass",
    "raise",
    "return",
    "try",
    "while",
    "with",
    "yield",
}


# Operators, ordered by descending length for longest-match scanning
OPERATORS = [
    # 3-char operators
    "<<=",
    ">>=",
    "**=",
    "//=",
    "...",
    "->",
    ":=",
    # 2-char operators
    "==",
    "!=",
    "<=",
    ">=",
    "<<",
    ">>",
    "**",
    "//",
    "+=",
    "-=",
    "*=",
    "/=",
    "%=",
    "&=",
    "|=",
    "^=",
]

# Single-char operators
SINGLE_CHAR_OPERATORS = set("+-*/%&|^~@=<>!")

# Punctuation
PUNCTUATION = {
    "(", ")", "{", "}", "[", "]",
    ",", ":", ";", ".",
}


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.length = len(source)
        self.pos = 0
        self.line = 1
        self.col = 1

        self.indents: List[int] = [0]
        self.open_brackets = 0
        self.at_line_start = True

        self.tokens: List[Token] = []

    def tokenize(self) -> List[Token]:
        while self.pos < self.length:
            if self.at_line_start and self.open_brackets == 0:
                if self._handle_indentation_and_blank_lines():
                    # A NEWLINE for a blank/comment-only line was emitted; continue.
                    continue

            c = self._peek()
            if c in " \t\r":
                self._advance()
                continue

            if c == "#":
                # Comment: consume until newline
                while self.pos < self.length and self._peek() not in "\n\r":
                    self._advance()
                # Do not emit a token here; newline will be handled in next iteration
                continue

            if c == "\n":
                # Newline handling
                self._advance_newline()
                if self.open_brackets == 0:
                    self._emit(TokenType.NEWLINE, None)
                    self.at_line_start = True
                else:
                    # Inside brackets: newline is ignored (implicit line-joining)
                    self.at_line_start = True
                continue

            # Not whitespace/comment/newline: we are no longer at line start
            self.at_line_start = False

            # Strings
            if c in ("'", '"'):
                self._scan_string()
                continue

            # Numbers (including leading . for floats like .5)
            if c.isdigit() or (c == "." and self._peek_next().isdigit()):
                self._scan_number()
                continue

            # Identifiers and keywords
            if c == "_" or c.isalpha():
                self._scan_identifier_or_keyword()
                continue

            # Operators or punctuation
            if self._scan_operator_or_punct():
                continue

            # Anything else is an error
            raise LexerError(f"Invalid character {c!r}", self.line, self.col)

        # End of input: emit a trailing NEWLINE (if last token isn't NEWLINE) and DEDENTs, then EOF
        if not self.tokens or self.tokens[-1].type != TokenType.NEWLINE:
            self._emit(TokenType.NEWLINE, None)

        while len(self.indents) > 1:
            self.indents.pop()
            self._emit(TokenType.DEDENT, None)

        self._emit(TokenType.EOF, None)
        return self.tokens

    # Internal helpers

    def _emit(self, ttype: TokenType, value: Optional[Union[str, int, float]]):
        self.tokens.append(Token(ttype, value, self.line, self.col))

    def _peek(self, n: int = 0) -> str:
        idx = self.pos + n
        if idx >= self.length:
            return "\0"
        return self.source[idx]

    def _peek_next(self) -> str:
        return self._peek(1)

    def _advance(self) -> str:
        ch = self._peek()
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return ch

    def _advance_newline(self):
        # Normalize CRLF and CR newlines to a single newline
        ch = self._peek()
        if ch == "\r":
            self._advance()
            if self._peek() == "\n":
                self._advance()
        else:
            self._advance()
        # After consuming newline, line/col have been updated by _advance()

    def _handle_indentation_and_blank_lines(self) -> bool:
        # Count indentation spaces (treat tab as 4 spaces)
        start_pos = self.pos
        start_col = self.col
        indent = 0
        while True:
            c = self._peek()
            if c == " ":
                indent += 1
                self._advance()
            elif c == "\t":
                indent += 4
                self._advance()
            else:
                break

        # If line is blank or comment-only, consume rest of line and emit NEWLINE
        c = self._peek()
        if c in ("\n", "\r"):
            self._advance_newline()
            self._emit(TokenType.NEWLINE, None)
            self.at_line_start = True
            return True
        if c == "#":
            while self.pos < self.length and self._peek() not in "\n\r":
                self._advance()
            if self._peek() in ("\n", "\r"):
                self._advance_newline()
            self._emit(TokenType.NEWLINE, None)
            self.at_line_start = True
            return True

        # Non-blank line: compare indentation levels
        current_indent = self.indents[-1]
        if indent > current_indent:
            self.indents.append(indent)
            self._emit(TokenType.INDENT, None)
        elif indent < current_indent:
            # Dedent to a matching level
            while len(self.indents) > 1 and indent < self.indents[-1]:
                self.indents.pop()
                self._emit(TokenType.DEDENT, None)
            if indent != self.indents[-1]:
                raise LexerError("Inconsistent dedent level", self.line, start_col)
        # else equal: nothing to emit

        self.at_line_start = False
        return False

    def _scan_identifier_or_keyword(self):
        start_line, start_col = self.line, self.col
        start = self.pos
        self._advance()  # first char
        while True:
            c = self._peek()
            if c == "_" or c.isalnum():
                self._advance()
            else:
                break
        text = self.source[start:self.pos]
        if text in KEYWORDS:
            self.tokens.append(Token(TokenType.KEYWORD, text, start_line, start_col))
        else:
            self.tokens.append(Token(TokenType.IDENTIFIER, text, start_line, start_col))

    def _scan_number(self):
        start_line, start_col = self.line, self.col
        start = self.pos

        def consume_digits():
            while self._peek().isdigit():
                self._advance()

        saw_digits = False
        # Integer part
        if self._peek().isdigit():
            consume_digits()
            saw_digits = True

        # Fractional part
        is_float = False
        if self._peek() == "." and self._peek_next().isdigit():
            is_float = True
            self._advance()  # consume '.'
            consume_digits()

        # Exponent part
        if self._peek() in "eE":
            is_float = True
            self._advance()  # e or E
            if self._peek() in "+-":
                self._advance()
            if not self._peek().isdigit():
                raise LexerError("Invalid float literal exponent", self.line, self.col)
            consume_digits()

        text = self.source[start:self.pos]
        if is_float:
            try:
                val = float(text)
            except ValueError:
                raise LexerError("Invalid float literal", start_line, start_col)
            self.tokens.append(Token(TokenType.FLOAT, val, start_line, start_col))
        else:
            try:
                val = int(text)
            except ValueError:
                raise LexerError("Invalid integer literal", start_line, start_col)
            self.tokens.append(Token(TokenType.INT, val, start_line, start_col))

    def _scan_string(self):
        quote = self._peek()
        start_line, start_col = self.line, self.col
        self._advance()  # consume opening quote

        is_triple = False
        if self._peek() == quote and self._peek_next() == quote:
            # Triple-quoted string
            is_triple = True
            self._advance()  # second quote
            self._advance()  # third quote

        chars: List[str] = []
        while True:
            c = self._peek()
            if c == "\0":
                raise LexerError("Unterminated string literal", start_line, start_col)
            if is_triple:
                if c == quote and self._peek(1) == quote and self._peek(2) == quote:
                    # End of triple-quoted string
                    self._advance()
                    self._advance()
                    self._advance()
                    break
                else:
                    ch = self._advance()
                    chars.append(ch)
                    continue
            else:
                if c == "\n" or c == "\r":
                    raise LexerError("Unterminated string literal", start_line, start_col)
                if c == "\\":
                    self._advance()  # consume backslash
                    esc = self._peek()
                    if esc == "\0":
                        raise LexerError("Unterminated string literal", start_line, start_col)
                    # Simple escape handling; keep as-is
                    chars.append(self._advance())
                    continue
                if c == quote:
                    self._advance()  # consume closing quote
                    break
                chars.append(self._advance())

        self.tokens.append(Token(TokenType.STRING, "".join(chars), start_line, start_col))

    def _scan_operator_or_punct(self) -> bool:
        # Try multi-char operators first
        for op in OPERATORS:
            if self._match(op):
                self._emit_operator(op)
                return True

        c = self._peek()
        # Punctuation that can also be part of multi-char ops: handle remaining here
        if c in PUNCTUATION:
            # Ellipsis '...' already handled above as operator; here only single '.' case remains
            if c in "([{":
                self.open_brackets += 1
            elif c in ")]}":
                if self.open_brackets > 0:
                    self.open_brackets -= 1
            self._advance()
            self._emit(TokenType.PUNCTUATION, c)
            return True

        # Single-char operators (not covered above)
        if c in SINGLE_CHAR_OPERATORS:
            self._advance()
            self._emit(TokenType.OP, c)
            return True

        return False

    def _emit_operator(self, op: str):
        # Adjust bracket count for operators that are actually punctuation-like
        if op == "...":
            # treated as punctuation-like but keep as operator
            pass
        self._emit(TokenType.OP, op)

    def _match(self, text: str) -> bool:
        end = self.pos + len(text)
        if self.source[self.pos:end] == text:
            # Special-case: distinguish ':=' operator from ':' punctuation
            # Already handled by ordering; just advance and emit
            self.pos = end
            # update line/col positions accordingly
            for ch in text:
                if ch == "\n":
                    self.line += 1
                    self.col = 1
                else:
                    self.col += 1
            return True
        return False

File: src/lexer.cpp
#include "pylite/lexer.hpp"

#include <algorithm>
#include <array>
#include <cctype>
#include <cstddef>
#include <cstdint>
#include <string>
#include <string_view>
#include <unordered_set>
#include <utility>

namespace pylite {
namespace {

std::string normalizeLineEndings(const std::string &input) {
    std::string result;
    result.reserve(input.size());

    for (std::size_t i = 0; i < input.size(); ++i) {
        char ch = input[i];
        if (ch == '\r') {
            if (i + 1 < input.size() && input[i + 1] == '\n') {
                ++i;
            }
            result.push_back('\n');
        } else {
            result.push_back(ch);
        }
    }

    return result;
}

const std::unordered_set<std::string> kKeywords = {
    "False", "None",  "True",    "and",   "as",     "assert", "async", "await", "break",  "class",
    "continue", "def",  "del",     "elif",  "else",   "except", "finally", "for",   "from",   "global",
    "if",     "import", "in",     "is",    "lambda", "nonlocal", "not",   "or",    "pass",   "raise",
    "return", "try",   "while",   "with",  "yield"};

const std::array<std::string_view, 22> kOperators = {
    "<<=", ">>=", "**=", "//=", "...", "->", ":=", "==", "!=", "<=", ">=", "<<", ">>",
    "**",  "//",  "+=",  "-=",  "*=",  "/=", "%=", "&=", "|=", "^="};

const std::unordered_set<char> kSingleCharOperators = {
    '+', '-', '*', '/', '%', '&', '|', '^', '~', '@', '=', '<', '>', '!',
};

const std::unordered_set<char> kPunctuation = {
    '(', ')', '[', ']', '{', '}', ',', ':', ';', '.',
};

char decodeEscape(char escape) {
    switch (escape) {
        case '\\':
            return '\\';
        case '\'':
            return '\'';
        case '"':
            return '"';
        case 'n':
            return '\n';
        case 'r':
            return '\r';
        case 't':
            return '\t';
        case '0':
            return '\0';
        default:
            return escape;
    }
}

bool isIdentifierStart(char ch) {
    return std::isalpha(static_cast<unsigned char>(ch)) || ch == '_';
}

bool isIdentifierPart(char ch) {
    return std::isalnum(static_cast<unsigned char>(ch)) || ch == '_';
}

TokenLiteral makeStringLiteral(std::string value) {
    return TokenLiteral(std::move(value));
}

TokenLiteral makeIntegerLiteral(const std::string &text, std::size_t line, std::size_t column) {
    try {
        std::size_t idx = 0;
        long long val = std::stoll(text, &idx, 10);
        if (idx != text.size()) {
            throw LexerError("Invalid integer literal", line, column);
        }
        return TokenLiteral(static_cast<std::int64_t>(val));
    } catch (const std::invalid_argument &) {
        throw LexerError("Invalid integer literal", line, column);
    } catch (const std::out_of_range &) {
        throw LexerError("Integer literal out of range", line, column);
    }
}

TokenLiteral makeFloatLiteral(const std::string &text, std::size_t line, std::size_t column) {
    try {
        std::size_t idx = 0;
        double val = std::stod(text, &idx);
        if (idx != text.size()) {
            throw LexerError("Invalid float literal", line, column);
        }
        return TokenLiteral(val);
    } catch (const std::invalid_argument &) {
        throw LexerError("Invalid float literal", line, column);
    } catch (const std::out_of_range &) {
        throw LexerError("Float literal out of range", line, column);
    }
}

class LexerImpl {
public:
    explicit LexerImpl(std::string source) : source_(std::move(source)) {}

    std::vector<Token> run() {
        while (pos_ < source_.size()) {
            if (atLineStart_ && openBrackets_ == 0) {
                if (handleIndentationAndBlankLines()) {
                    continue;
                }
            }

            char ch = peek();

            if (ch == '\0') {
                break;
            }

            if (ch == ' ' || ch == '\t' || ch == '\f') {
                advance();
                continue;
            }

            if (ch == '#') {
                skipComment();
                continue;
            }

            if (ch == '\n') {
                std::size_t tokenLine = line_;
                std::size_t tokenColumn = column_;
                advance();
                if (openBrackets_ == 0) {
                    emit(TokenType::Newline, "\n", TokenLiteral{}, tokenLine, tokenColumn);
                }
                atLineStart_ = true;
                continue;
            }

            atLineStart_ = false;

            if (ch == '\'' || ch == '"') {
                scanString();
                continue;
            }

            if (std::isdigit(static_cast<unsigned char>(ch)) ||
                (ch == '.' && std::isdigit(static_cast<unsigned char>(peek(1))))) {
                scanNumber();
                continue;
            }

            if (isIdentifierStart(ch)) {
                scanIdentifierOrKeyword();
                continue;
            }

            if (scanOperatorOrPunctuation()) {
                continue;
            }

            throw LexerError(std::string("Invalid character '") + ch + "'", line_, column_);
        }

        finalize();
        return std::move(tokens_);
    }

private:
    char peek(std::size_t offset = 0) const {
        if (pos_ + offset >= source_.size()) {
            return '\0';
        }
        return source_[pos_ + offset];
    }

    char advance() {
        if (pos_ >= source_.size()) {
            return '\0';
        }
        char ch = source_[pos_++];
        if (ch == '\n') {
            ++line_;
            column_ = 1;
        } else {
            ++column_;
        }
        return ch;
    }

    bool match(std::string_view text) {
        if (text.empty()) {
            return false;
        }
        if (pos_ + text.size() > source_.size()) {
            return false;
        }
        if (source_.compare(pos_, text.size(), text) != 0) {
            return false;
        }
        for (char ch : text) {
            (void)ch;
            advance();
        }
        return true;
    }

    void emit(TokenType type, std::string lexeme, TokenLiteral literal, std::size_t line, std::size_t column) {
        tokens_.push_back(Token{type, std::move(lexeme), std::move(literal), line, column});
    }

    void emitSimple(TokenType type, const std::string &lexeme, std::size_t line, std::size_t column) {
        emit(type, lexeme, TokenLiteral{}, line, column);
    }

    void emitSimple(TokenType type, char lexeme, std::size_t line, std::size_t column) {
        emit(type, std::string(1, lexeme), TokenLiteral{}, line, column);
    }

    bool handleIndentationAndBlankLines() {
        std::size_t indentLine = line_;
        std::size_t indentColumn = column_;
        std::size_t indentCount = 0;

        while (true) {
            char ch = peek();
            if (ch == ' ') {
                advance();
                ++indentCount;
            } else if (ch == '\t') {
                advance();
                indentCount += 4;
            } else {
                break;
            }
        }

        char ch = peek();
        if (ch == '\0') {
            return false;
        }

        if (ch == '\n') {
            std::size_t tokenLine = line_;
            std::size_t tokenColumn = column_;
            advance();
            emit(TokenType::Newline, "\n", TokenLiteral{}, tokenLine, tokenColumn);
            atLineStart_ = true;
            return true;
        }

        if (ch == '#') {
            skipComment();
            if (peek() == '\n') {
                std::size_t tokenLine = line_;
                std::size_t tokenColumn = column_;
                advance();
                emit(TokenType::Newline, "\n", TokenLiteral{}, tokenLine, tokenColumn);
                atLineStart_ = true;
                return true;
            }
            return false;
        }

        int currentIndent = indentStack_.back();
        int indentValue = static_cast<int>(indentCount);
        if (indentValue > currentIndent) {
            indentStack_.push_back(indentValue);
            emit(TokenType::Indent, "", TokenLiteral{}, indentLine, indentColumn);
        } else if (indentValue < currentIndent) {
            while (indentValue < indentStack_.back()) {
                indentStack_.pop_back();
                emit(TokenType::Dedent, "", TokenLiteral{}, indentLine, indentColumn);
            }
            if (indentValue != indentStack_.back()) {
                throw LexerError("Inconsistent dedent level", indentLine, indentColumn);
            }
        }

        atLineStart_ = false;
        return false;
    }

    void skipComment() {
        while (true) {
            char ch = peek();
            if (ch == '\0' || ch == '\n') {
                break;
            }
            advance();
        }
    }

    void scanIdentifierOrKeyword() {
        std::size_t tokenLine = line_;
        std::size_t tokenColumn = column_;
        std::size_t start = pos_;
        advance();
        while (isIdentifierPart(peek())) {
            advance();
        }
        std::string text = source_.substr(start, pos_ - start);
        if (kKeywords.find(text) != kKeywords.end()) {
            emit(TokenType::Keyword, text, makeStringLiteral(text), tokenLine, tokenColumn);
        } else {
            emit(TokenType::Identifier, text, makeStringLiteral(text), tokenLine, tokenColumn);
        }
    }

    void scanNumber() {
        std::size_t tokenLine = line_;
        std::size_t tokenColumn = column_;
        std::size_t start = pos_;
        bool sawDigits = false;

        auto consumeDigits = [this](bool &flag) {
            while (std::isdigit(static_cast<unsigned char>(peek()))) {
                advance();
                flag = true;
            }
        };

        consumeDigits(sawDigits);

        bool isFloat = false;

        if (peek() == '.' && std::isdigit(static_cast<unsigned char>(peek(1)))) {
            isFloat = true;
            advance();
            consumeDigits(sawDigits);
        }

        if (!sawDigits && peek() == '.' && std::isdigit(static_cast<unsigned char>(peek(1)))) {
            // Leading dot handled above, but ensure we mark as float
            isFloat = true;
        }

        if (peek() == 'e' || peek() == 'E') {
            isFloat = true;
            advance();
            if (peek() == '+' || peek() == '-') {
                advance();
            }
            if (!std::isdigit(static_cast<unsigned char>(peek()))) {
                throw LexerError("Invalid float literal exponent", line_, column_);
            }
            bool dummy = false;
            consumeDigits(dummy);
        }

        std::string text = source_.substr(start, pos_ - start);
        if (isFloat || text.find('.') != std::string::npos || text.find('e') != std::string::npos ||
            text.find('E') != std::string::npos) {
            emit(TokenType::Float, text, makeFloatLiteral(text, tokenLine, tokenColumn), tokenLine, tokenColumn);
        } else {
            emit(TokenType::Integer, text, makeIntegerLiteral(text, tokenLine, tokenColumn), tokenLine, tokenColumn);
        }
    }

    void scanString() {
        std::size_t tokenLine = line_;
        std::size_t tokenColumn = column_;
        std::size_t start = pos_;
        char quote = advance();
        bool isTriple = false;

        if (peek() == quote && peek(1) == quote) {
            isTriple = true;
            advance();
            advance();
        }

        std::string value;

        while (true) {
            char ch = peek();
            if (ch == '\0') {
                throw LexerError("Unterminated string literal", tokenLine, tokenColumn);
            }
            if (!isTriple && ch == '\n') {
                throw LexerError("Unterminated string literal", tokenLine, tokenColumn);
            }
            if (isTriple && ch == quote && peek(1) == quote && peek(2) == quote) {
                advance();
                advance();
                advance();
                break;
            }
            if (!isTriple && ch == quote) {
                advance();
                break;
            }
            if (ch == '\\') {
                advance();
                char esc = peek();
                if (esc == '\0') {
                    throw LexerError("Unterminated string literal", tokenLine, tokenColumn);
                }
                char decoded = decodeEscape(advance());
                value.push_back(decoded);
                continue;
            }
            value.push_back(advance());
        }

        std::string lexeme = source_.substr(start, pos_ - start);
        emit(TokenType::String, lexeme, makeStringLiteral(value), tokenLine, tokenColumn);
    }

    bool scanOperatorOrPunctuation() {
        std::size_t tokenLine = line_;
        std::size_t tokenColumn = column_;

        for (std::string_view op : kOperators) {
            if (match(op)) {
                emit(TokenType::Operator, std::string(op), TokenLiteral{}, tokenLine, tokenColumn);
                return true;
            }
        }

        char ch = peek();
        if (kPunctuation.find(ch) != kPunctuation.end()) {
            advance();
            if (ch == '(' || ch == '[' || ch == '{') {
                ++openBrackets_;
            } else if ((ch == ')' || ch == ']' || ch == '}') && openBrackets_ > 0) {
                --openBrackets_;
            }
            emitSimple(TokenType::Punctuation, ch, tokenLine, tokenColumn);
            return true;
        }

        if (kSingleCharOperators.find(ch) != kSingleCharOperators.end()) {
            advance();
            emitSimple(TokenType::Operator, ch, tokenLine, tokenColumn);
            return true;
        }

        return false;
    }

    void finalize() {
        if (tokens_.empty() || tokens_.back().type != TokenType::Newline) {
            emit(TokenType::Newline, "\n", TokenLiteral{}, line_, column_);
        }
        while (indentStack_.size() > 1) {
            indentStack_.pop_back();
            emit(TokenType::Dedent, "", TokenLiteral{}, line_, column_);
        }
        emit(TokenType::EndOfFile, "", TokenLiteral{}, line_, column_);
    }

private:
    std::string source_;
    std::size_t pos_ = 0;
    std::size_t line_ = 1;
    std::size_t column_ = 1;
    std::vector<int> indentStack_{0};
    int openBrackets_ = 0;
    bool atLineStart_ = true;
    std::vector<Token> tokens_;
};

}  // namespace

LexerError::LexerError(std::string message, std::size_t line, std::size_t column)
    : std::runtime_error(message + " at line " + std::to_string(line) + ", column " + std::to_string(column)),
      line_(line),
      column_(column) {}

std::vector<Token> Lexer::tokenize(const std::string &source) const {
    LexerImpl impl(normalizeLineEndings(source));
    return impl.run();
}

}  // namespace pylite
